infrastructure/README.md

```diff
@@ -0,0 +1,129 @@
+so we will use it to contain every app that we want to run.
+It doesn't matter that `Docker` isn't declarative since we don't care if the actual apps are.
+
+I know, maybe some of you wanted `lxce` or `nixOS`, I'm sorry.
+
+## Freight
+
+We will be using Kubernetes, since its gives us an easy way to run a an arbitrary amount of containers
+at once.
+
+## The App Repo
+
+Often a design can be gleaned from thinking of "what we would like to happen".
+So if I'm writing an app for a hackathon I would like it to be portable across all events,
+and in doing so I do not want to be responsible for placing this app in any specific event.
+
+So the steps to writing an app would be:
+
+1. Write your damn app
+2. Create a [Dockerfile](https://docs.docker.com/engine/reference/builder/) that runs your app.
+3. Create a `deployment.yaml` file which specifies what resources your app needs and
+   what other containers (if any) it depends on. This will be equivalent to the `spec.containers`
+   section of a [K8s deployment configuration](http://kubernetes.io/docs/user-guide/deployments/).
+
+Now when building or running the docker image the developer will have access to the entire event metadata
+(described later) in the form of JSON in an environment variable:
+
+```bash
+HACKATHON_METADATA='{"name":"HackGT","domain":"hack.gt", ...}'
+```
+
+This is helpful if you want to display the name of the event, etc. in your app.
+
+### Tests
+
+Every file that matches `/tests/Dockerfile` will spawn and run during testing.
+If you test different sections of your app, say one is called `websocket`, you can test
+the websocket capabilities of your app by writing tests in `websocket/tests/` then making
+a container to run the tests in.
+
+## The Event Repo
+
+When we know what event we are running these apps on we get to know a lot more,
+this is a simple repo that contains metadata about each event:
+
+```yaml
+name: "HackGT"
+times:
+  - start: 12345
+  - end: 12345
+  - submission: 12345
+homepage: "https://hack.gt/"
+domain: "hack.gt"
+apps:
+  hackgt/helpq:
+    - branch: master # this is default
+    - url: helpmewith.hack.gt
+  hackgt/eventbriter:
+    - domain: tickets
+  etc..
+```
+
+This metadata gets its own repository and is given a special name: `hackathon.yaml`.
+
+## The CI
+
+So far so simple I think. We've pushed off most of the responsibilities away from the people writing
+the apps: they only need to provide a container and what resources they want to use (ports, DBs, etc.).
+
+The CI therefore, by process of elimination, will do all the work for us. Specifically it will:
+
+ - Spawn an instance of any app that is under review through a pull request
+ - Run tests on all apps under review through a pull request
+ - Find all `hackathon.yaml` files and
+   - Create a namespace for that event
+   - Create a load balancer for that hackathon behind its `metadata.domain`
+   - Run all apps' tests
```

Comment:
Add: create a load balancer for each app to be able to do rolling upgrades and redundancy.